pest_meta/src/validator.rs

```diff
@@ -0,0 +1,535 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::collections::{HashMap, HashSet};
+
+use pest::Error;
+use pest::inputs::{Input, Span};
+use pest::iterators::Pairs;
+
+use quote::Ident;
+
+use super::parser::{PestRule, ParserExpr, ParserNode, ParserRule};
+
+pub fn validate_pairs<I: Input>(pairs: Pairs<PestRule, I>) -> Result<Vec<Ident>, Vec<Error<PestRule, I>>> {
+    let rust_keywords = hashset!{
+        "abstract", "alignof", "as", "become", "box", "break", "const", "continue", "crate", "do",
+        "else", "enum", "extern", "false", "final", "fn", "for", "if", "impl", "in", "let", "loop",
+        "macro", "match", "mod", "move", "mut", "offsetof", "override", "priv", "proc", "pure",
+        "pub", "ref", "return", "Self", "self", "sizeof", "static", "struct", "super", "trait",
+        "true", "type", "typeof", "unsafe", "unsized", "use", "virtual", "where", "while", "yield"
+    };
+    let pest_keywords = hashset!{"any", "eoi", "peek", "pop", "push", "soi"};
+    let predefined = hashset!{"any", "eoi", "peek", "pop", "soi"};
+
+    let definitions: Vec<_> = pairs.clone()
+                                   .filter(|pair| pair.as_rule() == PestRule::grammar_rule)
+                                   .map(|pair| {
+                                       pair.into_inner()
+                                           .next()
+                                           .unwrap()
+                                           .into_span()
+                                   }).collect();
+    let called_rules: Vec<_> = pairs.filter(|pair| pair.as_rule() == PestRule::grammar_rule)
+                                    .flat_map(|pair| {
+                                        pair.into_inner().flatten().skip(1).filter(|pair| {
+                                            pair.as_rule() == PestRule::identifier
+                                        }).map(|pair| {
+                                            pair.into_span()
+                                        })
+                                    }).collect();
+
+    let mut errors = vec![];
+
+    errors.extend(validate_rust_keywords(&definitions, &rust_keywords));
+    errors.extend(validate_pest_keywords(&definitions, &pest_keywords));
+    errors.extend(validate_already_defined(&definitions));
+    errors.extend(validate_undefined(&definitions, &called_rules, &predefined));
+
+    if errors.len() > 0 {
+        return Err(errors)
+    }
+
+    let definitions: HashSet<_> = definitions.iter().map(|span| span.as_str()).collect();
+    let called_rules: HashSet<_> = called_rules.iter().map(|span| span.as_str()).collect();
+
+    let defaults = called_rules.difference(&definitions);
+
+    Ok(defaults.into_iter().map(|string| Ident::new(*string)).collect())
+}
+
+pub fn validate_rust_keywords<'a, I: Input+'a, Definitions: IntoIterator<Item=&'a Span<I>>>(
```

Comment:
Sure thing! I tried [rustfmt](https://github.com/rust-lang-nursery/rustfmt), but had to revert because it had some very different opinions :smile: and I really wanted to reproduce the style of the other crates as much as possible. Is there any plan to use it at some point when it is stable (and consistent) enough?

Reply:
The issue with `rustfmt` is an [old one](https://github.com/pest-parser/pest/issues/148).

Reply:
Ah, whoops!  it was right there! Good thing rustfmt is open source.

Reply:
Yes, but, unfortunately, the hang bug seems to be pretty non-trivial.